test/e2e_node/restart_test.go

```diff
@@ -474,6 +474,73 @@ var _ = SIGDescribe("Restart", framework.WithSerial(), framework.WithSlow(), fra
 				return checkMirrorPodDisappear(ctx, f.ClientSet, pod.Name, pod.Namespace)
 			}, f.Timeouts.PodDelete, f.Timeouts.Poll).Should(gomega.BeNil())
 		})
+		// Regression test for an extended scenario for https://issues.k8s.io/123980
+		ginkgo.It("should evict running pods that do not meet the affinity after the kubelet restart", func(ctx context.Context) {
+			nodeLabelKey := "custom-label-key-required"
+			nodeLabelValueRequired := string(uuid.NewUUID())
+			podName := "affinity-pod" + string(uuid.NewUUID())
+			nodeName := getNodeName(ctx, f)
+			ginkgo.By(fmt.Sprintf("Adding node label for node (%v) to satisify pod (%v/%v) affinity", nodeName, f.Namespace.Name, podName))
+			e2enode.AddOrUpdateLabelOnNode(f.ClientSet, nodeName, nodeLabelKey, nodeLabelValueRequired)
+			pod := e2epod.MustMixinRestrictedPodSecurity(&v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      podName,
+					Namespace: f.Namespace.Name,
+				},
+				Spec: v1.PodSpec{
+					Affinity: &v1.Affinity{
+						NodeAffinity: &v1.NodeAffinity{
+							RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+								NodeSelectorTerms: []v1.NodeSelectorTerm{
+									{
+										MatchExpressions: []v1.NodeSelectorRequirement{
+											{
+												Key:      nodeLabelKey,
+												Operator: v1.NodeSelectorOpIn,
+												Values:   []string{nodeLabelValueRequired},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+					RestartPolicy: v1.RestartPolicyNever,
+					Containers: []v1.Container{
+						{
+							Name:  podName,
+							Image: imageutils.GetPauseImageName(),
+						},
+					},
+				},
+			})
+
+			// Create the pod bound to the node. It will start, but will be rejected after kubelet restart.
+			ginkgo.By(fmt.Sprintf("Creating a pod (%s/%s)", f.Namespace.Name, podName))
+			e2epod.NewPodClient(f).Create(ctx, pod)
+
+			ginkgo.By(fmt.Sprintf("Waiting for the pod (%s/%s) to be running", f.Namespace.Name, pod.Name))
+			err := e2epod.WaitForPodNameRunningInNamespace(ctx, f.ClientSet, pod.Name, f.Namespace.Name)
+			framework.ExpectNoError(err, "Failed to await for the pod to be running: (%v/%v)", f.Namespace.Name, pod.Name)
+
+			// Remove node label
+			e2enode.RemoveLabelOffNode(f.ClientSet, nodeName, nodeLabelKey)
+
+			ginkgo.By("Restart the kubelet")
+			restartKubelet(true)
+			gomega.Eventually(ctx, func() bool {
+				return kubeletHealthCheck(kubeletHealthCheckURL)
+			}, recoverTimeout, f.Timeouts.Poll).
+				Should(gomega.BeTrueBecause("kubelet should be healthy after restart"))
+
+			// Pod should be terminated, maybe not immediately, should a few seconds let the kubelet to kill the pod
+			gomega.Eventually(ctx, func() bool {
+				pod, err = e2epod.NewPodClient(f).Get(ctx, podName, metav1.GetOptions{})
+				framework.ExpectNoError(err)
+				// pod is in a final state
+				return pod.Status.Phase == v1.PodFailed || pod.Status.Phase == v1.PodSucceeded
```

Comment:
When the kubelet evicts a Pod, even the same Pod can end up in either the `PodFailed` or `PodSucceeded` state (I'm not sure why, but I think I've seen this happen before). Is that why you're considering both cases here?

Reply:
I seem to have encountered this before, i think as long as the pod shows the final status, it is expected, although `PodSucceeded` is a bit strange..

Reply:
Let's only check for what is expected - we expect that the pod admission will fail. Waiting for a miracle is bad, we need to test exactly the behavior we expect and have an explanation if the behavior is different

Reply:
@SergeyKanzhelev @toVersus ,this e2e pod Status.Phase should be `PodSucceeded`.

The following are the behaviors of pods after kubelet restarts:
>1. kubelet `canAdmitPod` reject pod by reason `Pod admission denied by nodeAffinity`
>2. kubelet kill container

The final state of the pod is related to the kill container. If an error occurs in the preStop of the container or the exitCode is not 0, the phase is `PodFailed`. If the exitCode of the StopContainer is 0, the phase is `PodSucceeded`.

So the status of this test case should always be `PodSucceeded`

Reply:
wow, this is unexpected. Would you mind adding this explanation to the comments so it will be more obvious why we expect it to succeed.